Add tutorial "Using Stamped datatypes with tf2_ros::MessageFilter"

kennywangaigraphx opened the pull request.

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped_datatypes-with-tf2_ros-MessageFilter.rst:
```diff
@@ -0,0 +1,564 @@
+1 Write the broadcaster node of PointStamped messages
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+For this tutorial we will set up a demo application which has a node(in Python) to broadcast the ``PointStamped`` messages of ``turtle3``.
```
I think we can add a space before the bracket

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped_datatypes-with-tf2_ros-MessageFilter.rst:
```diff
@@ -0,0 +1,564 @@
+
+First, let's create the source file.
+
+Go to the ``learning_tf2_py`` package we created in the previous tutorial.
```
May be add a link here?

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped_datatypes-with-tf2_ros-MessageFilter.rst:
```diff
@@ -0,0 +1,564 @@
+
+        .. code-block:: console
+
+            wget https://raw.githubusercontent.com/ros/geometry_tutorials/ros2/turtle_tf2_py/turtle_tf2_py/turtle_tf2_message_broadcaster.py
```
This file does not exist in the https://github.com/ros/geometry_tutorials yet. Do mind to open a PR to add that file there?

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped_datatypes-with-tf2_ros-MessageFilter.rst:
```diff
@@ -0,0 +1,564 @@
+
+Open the file using your preferred text editor.
+
+.. code-block::python
```
```suggestion
.. code-block:: python

```

it is needed to add this space as the code did not appear in the built html page without it

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped_datatypes-with-tf2_ros-MessageFilter.rst:
```diff
@@ -0,0 +1,564 @@
+        
+       def handle_turtle_pose(self,msg):
+           vel_msg = Twist()
+           vel_msg.linear.x =1.0
```
```suggestion
           vel_msg.linear.x = 1.0

```

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped_datatypes-with-tf2_ros-MessageFilter.rst:
```diff
@@ -0,0 +1,564 @@
+1.1 Examine the code
+~~~~~~~~~~~~~~~~~~~~~
+
+Now, let's take a look at the code.
```
```suggestion
Now let's take a look at the code.

```

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped_datatypes-with-tf2_ros-MessageFilter.rst:
```diff
@@ -0,0 +1,564 @@
+~~~~~~~~~~~~~~~~~~~~~
+
+Now, let's take a look at the code.
+Firstly, we spawn the ``turtle3`` by asynchronously calling the ``Spawn`` service of ``turtlesim``, and initialize its position in (4, 2, 0).
```
I think the correct preposition here would be: at (4, 2, 0).

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped_datatypes-with-tf2_ros-MessageFilter.rst:
```diff
@@ -0,0 +1,564 @@
+
+.. code-block:: python
+
+    from launch import Laun~~chDescription
```
looks like a typo here

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped_datatypes-with-tf2_ros-MessageFilter.rst:
```diff
@@ -0,0 +1,564 @@
+2.1 Examine the code
+~~~~~~~~~~~~~~~~~~~~~
+
+Firstly, the includes. You must include the tf2_ros::MessageFilter headers from the tf2_ros package. As well as the previously used tf2 and ros2 related headers.
```
```suggestion
Firstly, you must include the ``tf2_ros::MessageFilter`` headers from the ``tf2_ros`` package. As well as the previously used ``tf2`` and ``ros2`` related headers.

```

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped_datatypes-with-tf2_ros-MessageFilter.rst:
```diff
@@ -0,0 +1,564 @@
+   #include <message_filters/subscriber.h>
+   
+
+Secondly, the persistent data. There need to be persistent instances of tf2_ros::Buffer, tf2_ros::TransformListener and tf2_ros::MessageFilter.
```
```suggestion
Secondly, the persistent data. There need to be persistent instances of ``tf2_ros::Buffer``, ``tf2_ros::TransformListener`` and ``tf2_ros::MessageFilter``.

```

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped_datatypes-with-tf2_ros-MessageFilter.rst:
```diff
@@ -0,0 +1,564 @@
+     tf2_filter_->registerCallback(&PoseDrawer::msgCallback, this);
+   }  
+
+And lastly, the callback method. Once the data is ready, just call tf2_buffer_->transform and print to screen for the tutorial.
```
It is always best to wrap code snippets. It is better for visibility and for the complication. For example, this code snippet created a link in the text that is not getting anywhere.

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped_datatypes-with-tf2_ros-MessageFilter.rst:
```diff
@@ -0,0 +1,564 @@
+   std::shared_ptr<tf2_ros::MessageFilter<geometry_msgs::msg::PointStamped>> tf2_filter_;
+
+
+Thirdly, the constructor. When starting up the ros message_filters::Subscriber must be initialized with the topic. And the tf2_ros::MessageFilter must be initialized with that Subscriber object. The other arguments of note in the MessageFilter constructor are the target_frame and callback function. The target frame is the frame into which it will make sure canTransform will succeed. And the callb ...
```
```suggestion
Thirdly, the constructor. When starting up the ROS 2 ``message_filters::Subscriber`` must be initialized with the topic. And the ``tf2_ros::MessageFilter`` must be initialized with that ``Subscriber`` object. The other arguments of note in the ``MessageFilter`` constructor are the ``target_frame`` and callback function. The target frame is the frame into which it will make sure ``canTransform`` will succeed. And the callback function is the function which will be called when the data is ready.

```

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped_datatypes-with-tf2_ros-MessageFilter.rst:
```diff
@@ -0,0 +1,564 @@
+2.2 Build the package
+~~~~~~~~~~~~~~~~~~~~~~
+
+Before rebuild the package ``learning_tf2_cpp``, please add two another dependencies in the ``package.xml`` file of this package:
```
maybe change rebuild to building ?

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped_datatypes-with-tf2_ros-MessageFilter.rst:
```diff
@@ -0,0 +1,564 @@
+
+   ros2 launch learning_tf2_py turtle_tf2_sensor_message.launch.py
+
+This will bring up the ``turtlesim`` window, where have two turtles, ``turtle3`` is moving along a circle, while ``turtle1`` isn't moving at first. But you can run the ``turtle_teleop_key`` node in another terminal to drive ``turtle1`` to move:
```
```suggestion
This will bring up the ``turtlesim`` window with two turtles, where ``turtle3`` is moving along a circle, while ``turtle1`` isn't moving at first. But you can run the ``turtle_teleop_key`` node in another terminal to drive ``turtle1``:

```

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped_datatypes-with-tf2_ros-MessageFilter.rst:
```diff
@@ -0,0 +1,564 @@
+    self.pub.publish(ps)
+
+1.2 Write the launch file
+~~~~~~~~~~~~~~~~~~~~~~~~~~
```
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~

```

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped_datatypes-with-tf2_ros-MessageFilter.rst:
```diff
@@ -0,0 +1,564 @@
+
+
+1.3 Add an entry point and build the pakcage
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped_datatypes-with-tf2_ros-MessageFilter.rst:
```diff
@@ -0,0 +1,564 @@
+         
+
+Running and results
+--------------------
```
```suggestion
-------------------

```

kurshakuz reviewed (changes requested):
You did a great job, @kennywangaigraphx! Overall, very well structured tutorial, with only some minor moments to resolve. Firstly, the links to the code in `geometry_tutorials` that are included does not exist yet. Therefore, I suggest to open a PR there and add your nodes and launch files. The code will be reviewed there too.

About the text part of the tutorial, as I said everything is great, and just requires a couple of updates. One thing is that it is always better to wrap classes, messages, and all inline code in text in the code snippets. 

Additionally, update the tf2 main page to include a link to that tutorial.

I also think that it is best to rename the tutorial file name to make it consistent with the rest. This camel case look should work I guess:  
`Using-Stamped_datatypes-with-tf2_ros-MessageFilter.rst ` -> `Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst `

Other suggestions are minor, and you can take a look at them below.

clalancette commented on source/Tutorials/Tf2/Using-Stamped_datatypes-with-tf2_ros-MessageFilter.rst:
```diff
@@ -0,0 +1,564 @@
+
+    * cameras, both mono and stereo
+    * laser scans
+Suppose that a new turtle named ``turtle3`` is created and it doesn't have good odometry, but there is an overhead camera tracking its position and publishing it as a `PointStamped <https://wiki.ros.org/PointStamped>`_ message in relation to the ``world`` frame.
```
As a general comment, we should not have links out to https://wiki.ros.org .  That is the wiki for ROS 1, and will only be for ROS 1.  Anything for ROS 2 should be on https://docs.ros.org .

That said, we don't have API documentation currently being generated for the `.msg` files in Rolling (as @kurshakuz pointed out).  So let's just not put these links in for now, and we can always add them later on when we get message documentation.

kurshakuz commented:
This PR requires https://github.com/ros/geometry_tutorials/pull/62 to be merged first.

kennywangaigraphx commented:
Thank you very much for your review, @kurshakuz  and @clalancette . Firstly, I created a PR just now in "geometry_tutorials", https://github.com/ros/geometry_tutorials/pull/62 ; Secondly, I updated the Tf2-Main.rst  to include a link to this tutorial; Thirdly, I wraped my codes in snippets, but there still have link in some variables in the code, but no link in the text now; The fourth, I modified the tutorial according to all the minor suggestions of @kurshakuz . 
Sorry, but I don't know how to fill it here, but you can review the changes in "Files changed".

ahcorde commented on source/Tutorials/Tf2/Tf2-Main.rst:
```diff
@@ -1,71 +1,78 @@
-.. _Tf2Main:
```
why all these changes are happening ? if there is no change at all ?

kurshakuz replied:
I had the same issue

kurshakuz replied:
I had the same issue that was solved in:
https://github.com/ros2/ros2_documentation/pull/1923/commits/692519cbc44e7bd5ab7852727e91666bcc405e9b

The problem is that some old MacOS used CR end of line sequence. He will need to change that to CR LF:
https://stackoverflow.com/questions/1552749/difference-between-cr-lf-lf-and-cr-line-break-types

@kennywangaigraphx can you change them using your text editor? As an example, I used VSCode to do this following that suggestion:
https://stackoverflow.com/a/39532890

kennywangaigraphx replied:
@kurshakuz Ok, I'll  change them to CRLF.

kennywangaigraphx replied:
@ahcorde I checked linters in https://github.com/ros/geometry_tutorials/pull/62 , and I'll change codes in this PR accordingly.

kennywangaigraphx replied:
@ahcorde @kurshakuz Finally it seems that it passed the test.

ahcorde reviewed (changes requested):
Can you check linters ?

kennywangaigraphx commented:
@kurshakuz There still has conflicts in the Tf2-Main.rst file, would you please guide me how to resolve it?

kurshakuz commented:
As I see it now, the end of line sequence is still not CR LF, cause the last commit still wants to overwrite whole file. Try fixing it again

kennywangaigraphx commented:
@kurshakuz  I tried to change the end of line sequence to CRLF, but it will overwrite the whole file. Then I followed your commit to change the end of line sequence to LF, and it will not overwrite the whole file now, but there have two conflicts: each place  where I add Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter will have conflict.
I was confused by your commit https://github.com/ros2/ros2_documentation/commit/692519cbc44e7bd5ab7852727e91666bcc405e9b, because its title is "change End Of Line sequence from CRLF to LF ", so I want to know which one really will work, CRLF or LF?

kurshakuz commented:
> @kurshakuz I tried to change the end of line sequence to CRLF, but it will overwrite the whole file. Then I followed your commit to change the end of line sequence to LF, and it will not overwrite the whole file now, but there have two conflicts: each place where I add Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter will have conflict.
> I was confused by your commit [692519c](https://github.com/ros2/ros2_documentation/commit/692519cbc44e7bd5ab7852727e91666bcc405e9b), because its title is "change End Of Line sequence from CRLF to LF ", so I want to know which one really will work, CRLF or LF?

you need to use LF

kurshakuz commented on source/Tutorials/Tf2/Tf2-Main.rst:
```diff
@@ -69,3 +71,16 @@ Learning tf2
+
+   This tutorial teaches you about a systematic approach for debugging tf2 related problems.
+
+Using sensor messages with tf2
```
Now the EOL sequence is correct

kurshakuz replied:
Can you also check if the same format is used in the tutorial page?

kennywangaigraphx replied:
Ok, then the LF will work, I think.

kennywangaigraphx replied:
But it still have two conflicts in the Tf2-Main.rst file.

kurshakuz replied:
You need to merge those updates from the rolling branch

```
git remote add ros2 https://github.com/ros2/ros2_documentation/
git fetch ros2
git merge ros2/rolling
git push origin rolling
```

kennywangaigraphx replied:
My files are in the MessageFilter branch other than rolling branch,  what should the commands be? Thank you.

kennywangaigraphx replied:
@kurshakuz My files are in the MessageFilter branch other than rolling branch, what should the commands be?  Maybe the last command should be "git push origin MessageFilter"? or just the same as you told me? Thank you.

kennywangaigraphx replied:
@kurshakuz I push the origin MessageFilter branch, and there has some whitespace errors in the .rst file , and I corrected them and commited it just now.

kurshakuz commented on source/Tutorials/Tf2/Tf2-Main.rst:
```diff
@@ -27,7 +27,7 @@ If you want to learn both C++ and Python, you should go through the tutorials on
    Learning-About-Tf2-And-Time-Cpp
    Time-Travel-With-Tf2-Py
    Time-Travel-With-Tf2-Cpp
-   Debugging-Tf2-Problems
```
You don't need to overwrite this part, just add your tutorial one step below

kurshakuz commented on source/Tutorials/Tf2/Tf2-Main.rst:
```diff
@@ -68,9 +68,9 @@ Learning tf2
 
    This tutorial teaches you about advanced time travel features of tf2.
 
-Debugging tf2
```
Same as above

kennywangaigraphx replied:
@kurshakuz Sorry for that, I fixed it just now.

kennywangaigraphx replied:
@kurshakuz @ahcorde All checks have passed! So what should I do next?

kurshakuz replied:
@kennywangaigraphx very good! Now we will need to merge your PR in `geometry_tutorials` first

kennywangaigraphx replied:
@kurshakuz I got it, thank you very much for your help.

clalancette commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,609 @@
+
+``turtle1`` wants to know where ``turtle3`` is compared to itself.
+
+To do this ``turtle1`` must listen to the topic where ``turtle3``'s pose is being published, wait until transforms into the desired frame are ready, and then do it's operations. To make this easier the ``tf2_ros::MessageFilter`` is very useful. The ``tf2_ros::MessageFilter`` will take a subscription to any ROS 2 message with a header and cache it until it is possible to transform it into the targ ...
```
Here and elsewhere in this document, please do one sentence per-line.  That just makes it easier to make edits in the future.

kennywangaigraphx replied:
Yes, I did it just now, and now in the file each line only has one sentence.

clalancette commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,609 @@
+
+        .. code-block:: console
+
+                curl -sk https://raw.githubusercontent.com/ros/geometry_tutorials/ros2/turtle_tf2_py/turtle_tf2_py/turtle_tf2_message_broadcaster.py -o turtle_tf2_message_broadcaster.py
```
Please make the indentation consistent with above, that is, get rid of 4 spaces here.  Same for the second curl line down below.

kennywangaigraphx replied:
Ok, I checked and modified all the indentation consistent in my tutorial file.

kurshakuz commented on source/Tutorials/Tf2/Tf2-Main.rst:
```diff
@@ -74,3 +75,10 @@ Debugging tf2
+
+#. :ref:`Using stamped datatypes with tf2_ros::MessageFilter <UsingStampedDatatypesWithTf2RosMessageFilter>`.
+
+   This tutorial teaches you how to use tf2_ros::MessageFilter to process stamped datatypes.
```
```suggestion
   This tutorial teaches you how to use ``tf2_ros::MessageFilter`` to process stamped datatypes.
```

kennywangaigraphx replied:
I did that changing.

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,668 @@
+Using stamped datatypes with tf2_ros::MessageFilter
+===================================================
+
+**Goal:** Learn how to use tf2_ros::MessageFilter to process stamped datatypes.
```
```suggestion
**Goal:** Learn how to use ``tf2_ros::MessageFilter`` to process stamped datatypes.
```

kennywangaigraphx replied:
I did it just now.

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,668 @@
+
+``turtle1`` wants to know where ``turtle3`` is compared to itself.
+
+To do this ``turtle1`` must listen to the topic where ``turtle3``'s pose is being published, wait until transforms into the desired frame are ready, and then do it's operations.
```
```suggestion
To do this ``turtle1`` must listen to the topic where ``turtle3``'s pose is being published, wait until transforms into the desired frame are ready, and then do its operations.
```

kennywangaigraphx replied:
Yes, you're right, it's a syntax error, I changed it.

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,668 @@
+``turtle1`` wants to know where ``turtle3`` is compared to itself.
+
+To do this ``turtle1`` must listen to the topic where ``turtle3``'s pose is being published, wait until transforms into the desired frame are ready, and then do it's operations.
+To make this easier the ``tf2_ros::MessageFilter`` is very useful. The ``tf2_ros::MessageFilter`` will take a subscription to any ROS 2 message with a header and cache it until it is possible to transform it into the target frame.
```
Do you think we can paraphrase that first sentence? I think it would be best to clarify what is it making easier and how

kennywangaigraphx replied:
Yes, I think the "this" in the "To make this easier" sentence refers to the sentence "To do this "turtle1" must listen to ...", so I merged these two paragraphs into one, and I think that can clarify what is it making easier and how.

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,668 @@
+1 Write the broadcaster node of PointStamped messages
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+For this tutorial we will set up a demo application which has a node (in Python) to broadcast the ``PointStamped`` messages of ``turtle3``.
```
I think adding that we are publishing a ``PointStamped`` **position** messages of ``turtle3`` will make it clear

kennywangaigraphx replied:
I changed that.

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,668 @@
+   # distributed under the License is distributed on an "AS IS" BASIS,
+   # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   # See the License for the specific language governing permissions and
+   # limitations under the License.
```
Usually we omit copyright notices in code snippets

kennywangaigraphx replied:
Ok, I deleted all compyright notices in the code snippets.

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,668 @@
+   self.sub = self.create_subscription(Pose, '/turtle3/pose', self.handle_turtle_pose, 10)
+   self.pub = self.create_publisher(PointStamped, '/turtle3/turtle_point_stamped', 10)
+
+Finally, in the callback function ``handle_turtle_pose``, we initialize the ``Twist`` messages of ``turtle3`` and publish them, which will make the ``turtle3`` moving along a circle, then we fill up the ``PointStamped`` messages of ``turtle3`` with incoming ``Pose`` messages of itself and publish it.
```
```suggestion
Finally, in the callback function ``handle_turtle_pose``, we initialize the ``Twist`` messages of ``turtle3`` and publish them, which will make the ``turtle3`` move along a circle. Then we fill up the ``PointStamped`` messages of ``turtle3`` with incoming ``Pose`` messages and publish them.
```

kennywangaigraphx replied:
You're very careful to help me find these errors. I changed them, thank you.

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,668 @@
+   # distributed under the License is distributed on an "AS IS" BASIS,
+   # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   # See the License for the specific language governing permissions and
+   # limitations under the License.
```
same for this copyright

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,668 @@
+.. code-block:: python
+
+   'console_scripts': [
+       'turtle_tf2_msg_broadcaster = learning_tf2_py.turtle_tf2_message_broadcaster:main',
```
```suggestion
       ...
       'turtle_tf2_msg_broadcaster = learning_tf2_py.turtle_tf2_message_broadcaster:main',
```
I think it would be better to add three dots here to indicate that we are adding that node and not removing previous ones

kennywangaigraphx replied:
Yes, you're right. I added three dots here.

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,668 @@
+           ),
+           Node(
+               package='turtle_tf2_py',
+               executable='turtle_tf2_message_broadcaster',
```
Is this correct that the executable name here  and on line 289 are different? First is `turtle_tf2_message_broadcaster` and the second is `turtle_tf2_msg_broadcaster`. I think you need to update one in the line 289

kennywangaigraphx replied:
You're right, I modified "turtle_tf2_msg_broadercaster" in line 289 to "turtle_tf2_message_broadcaster".

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,668 @@
+   }
+
+
+You can download this code directly from file ``turtle_tf2_message_filter.cpp`` in the ``src`` subdirectory of the ``learning_tf2_cpp`` :ref:`package <WritingATf2StaticBroadcasterCpp>`.
```
I think it would be better to add wget links in a similar manner before the code itself and remove that part

kennywangaigraphx replied:
Ok, I added wget links before the code itself.

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,668 @@
+
+
+Thirdly, the constructor. When starting up the ROS 2 ``message_filters::Subscriber`` must be initialized with the topic. And the ``tf2_ros::MessageFilter`` must be initialized with that ``Subscriber`` object.
+The other arguments of note in the ``MessageFilter`` constructor are the ``target_frame`` and callback function. The target frame is the frame into which it will make sure ``canTransform`` will succeed. And the callback function is the function which will be called when the data is ready.
```
```suggestion
The other arguments of note in the ``MessageFilter`` constructor are the ``target_frame`` and the callback function. The target frame is the frame into which it will make sure ``canTransform`` will succeed. And the callback function is the function that will be called when the data is ready.
```

kennywangaigraphx replied:
Yes, my english is not good, so I actually don't know what differences between "which" and "that" when they are used in clause. Anyway, I changed it as your suggestion.

kurshakuz replied:
Mine is neither, don't worry 😂
We are all learning in the process

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,668 @@
+   }
+
+
+And lastly, the callback method. Once the data is ready, just call ``tf2_buffer_->transform`` and print to screen for the tutorial.
```
We can rephrase that as "print output to the console"

kennywangaigraphx replied:
I changed that, too.

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,668 @@
+Summary
+-------
+
+In this tutorial you learned how to use sensor data/messages in tf2. Specifically speaking, you learned how to publish PointStamped messages on a topic, and how to listen to the topic and transform the frame of PointStamped messages with tf2_ros::MessageFilter.
```
```suggestion
In this tutorial you learned how to use sensor data/messages in tf2. Specifically speaking, you learned how to publish ``PointStamped`` messages on a topic, and how to listen to the topic and transform the frame of ``PointStamped`` messages with ``tf2_ros::MessageFilter``.
```

kennywangaigraphx replied:
I did it.

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,668 @@
+
+   ros2 run turtlesim turtle_teleop_key
+
+.. image:: turtlesim_messagefilter.png
```
Can you crop the image to remove that window header? So that it matches all other pictures in tf2 tutorials

kennywangaigraphx replied:
Ok, I cropped the image just now.

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,668 @@
+
+**Goal:** Learn how to use tf2_ros::MessageFilter to process stamped datatypes.
+
+**Tutorial level:** Intermediate
```
You can mark this tutorial as Advanced I think

kennywangaigraphx replied:
Awesome, I think it's Advanced, so I changed that.

kennywangaigraphx replied:
And I move the image to the images sub-folder as you did, and changed the image path accordingly in the rst file.

kurshakuz reviewed (changes requested):
I believe that these are the last needed updates and we are ready to go. Great tutorial!

kennywangaigraphx commented:
@kurshakuz  Ok, I think I tried my best to change all the things you suggested. Thank you very much and @ahcorde  and @clalancette .

kurshakuz commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,656 @@
+
+Now, to get the streaming ``PointStamped`` data of ``turtle3`` in the frame of ``turtle1`` reliably, we will create the source file at first.
+
+Go to the ``learning_tf2_cpp`` :ref:`package <WritingATf2StaticBroadcasterCpp>` we created in the previous tutorial.Inside the ``src/learning_tf2_cpp/src`` directory download file ``turtle_tf2_message_filter.cpp`` by entering the following command:
```
```suggestion
Go to the ``learning_tf2_cpp`` :ref:`package <WritingATf2StaticBroadcasterCpp>` we created in the previous tutorial.
Inside the ``src/learning_tf2_cpp/src`` directory download file ``turtle_tf2_message_filter.cpp`` by entering the following command:
```

kennywangaigraphx replied:
Ok, I changed that line into two paragraphs.

kurshakuz reviewed (approved):
Looks good to me! I added one last nitpick

kennywangaigraphx commented:
@kurshakuz  @ahcorde  @clalancette  Do I need to do anything else for this PR to merge?

clalancette commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,657 @@
+
+``turtle1`` wants to know where ``turtle3`` is compared to itself.
+
+To do this ``turtle1`` must listen to the topic where ``turtle3``'s pose is being published, wait until transforms into the desired frame are ready, and then do its operations. To make this easier the ``tf2_ros::MessageFilter`` is very useful. The ``tf2_ros::MessageFilter`` will take a subscription to any ROS 2 message with a header and cache it until it is possible to transform it into the targe ...
```
```suggestion
To do this ``turtle1`` must listen to the topic where ``turtle3``'s pose is being published, wait until transforms into the desired frame are ready, and then do its operations.
To make this easier the ``tf2_ros::MessageFilter`` is very useful.
The ``tf2_ros::MessageFilter`` will take a subscription to any ROS 2 message with a header and cache it until it is possible to transform it into the target frame.
```

kennywangaigraphx replied:
Ok, I split this paragraph into 3 paragraphs.

clalancette commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,657 @@
+
+To do this ``turtle1`` must listen to the topic where ``turtle3``'s pose is being published, wait until transforms into the desired frame are ready, and then do its operations. To make this easier the ``tf2_ros::MessageFilter`` is very useful. The ``tf2_ros::MessageFilter`` will take a subscription to any ROS 2 message with a header and cache it until it is possible to transform it into the targe ...
+
+Setting the example
```
```suggestion
Setting up the example
```

kennywangaigraphx replied:
Yes, you're right, I should use "setting up" other than "setting". I changed that.

clalancette commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,657 @@
+   self.sub = self.create_subscription(Pose, '/turtle3/pose', self.handle_turtle_pose, 10)
+   self.pub = self.create_publisher(PointStamped, '/turtle3/turtle_point_stamped', 10)
+
+Finally, in the callback function ``handle_turtle_pose``, we initialize the ``Twist`` messages of ``turtle3`` and publish them, which will make the ``turtle3`` move along a circle. Then we fill up the ``PointStamped`` messages of ``turtle3`` with incoming ``Pose`` messages and publish them.
```
```suggestion
Finally, in the callback function ``handle_turtle_pose``, we initialize the ``Twist`` messages of ``turtle3`` and publish them, which will make the ``turtle3`` move along a circle.
Then we fill up the ``PointStamped`` messages of ``turtle3`` with incoming ``Pose`` messages and publish them.
```

kennywangaigraphx replied:
I split it into 2 paragraphs.

clalancette commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,657 @@
+2 Writing the message filter/listener node
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Now, to get the streaming ``PointStamped`` data of ``turtle3`` in the frame of ``turtle1`` reliably, we will create the source file at first.
```
I'm not quite sure what this sentence is saying.  Is it that we are now creating the source file for `PointStamped` data, or something else?  Clarification here would be useful.

kennywangaigraphx replied:
Yes, I clarified as following:
![image](https://user-images.githubusercontent.com/89616578/135030243-d21f8ffb-b150-4e88-b52e-30dd9e861be3.png)

clalancette commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,657 @@
+2.1 Examine the code
+~~~~~~~~~~~~~~~~~~~~
+
+Firstly, you must include the ``tf2_ros::MessageFilter`` headers from the ``tf2_ros`` package. As well as the previously used ``tf2`` and ``ros2`` related headers.
```
```suggestion
First, you must include the ``tf2_ros::MessageFilter`` headers from the ``tf2_ros`` package, as well as the previously used ``tf2`` and ``ros2`` related headers.
```

kennywangaigraphx replied:
I changed these errors.

clalancette commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,657 @@
+~~~~~~~~~~~~~~~~~~~~
+
+Now let's take a look at the code.
+Firstly, in the ``on_timer`` callback function, we spawn the ``turtle3`` by asynchronously calling the ``Spawn`` service of ``turtlesim``, and initialize its position at (4, 2, 0), when the turtle spawning service is ready.
```
```suggestion
First, in the ``on_timer`` callback function, we spawn the ``turtle3`` by asynchronously calling the ``Spawn`` service of ``turtlesim``, and initialize its position at (4, 2, 0), when the turtle spawning service is ready.
```

kennywangaigraphx replied:
Ok, the same error, I modified it.

clalancette commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,657 @@
+   #endif
+
+
+Secondly, the persistent data. There need to be persistent instances of ``tf2_ros::Buffer``, ``tf2_ros::TransformListener`` and ``tf2_ros::MessageFilter``.
```
```suggestion
Second, there needs to be persistent instances of ``tf2_ros::Buffer``, ``tf2_ros::TransformListener`` and ``tf2_ros::MessageFilter``.
```

kennywangaigraphx replied:
I changed it.

clalancette commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,657 @@
+   std::shared_ptr<tf2_ros::MessageFilter<geometry_msgs::msg::PointStamped>> tf2_filter_;
+
+
+Thirdly, the constructor. When starting up the ROS 2 ``message_filters::Subscriber`` must be initialized with the topic. And the ``tf2_ros::MessageFilter`` must be initialized with that ``Subscriber`` object.
```
```suggestion
Third, the ROS 2 ``message_filters::Subscriber`` must be initialized with the topic.
And the ``tf2_ros::MessageFilter`` must be initialized with that ``Subscriber`` object.
```

kennywangaigraphx replied:
I corrected it, and split this paragraph into two.

clalancette commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,657 @@
+
+
+Thirdly, the constructor. When starting up the ROS 2 ``message_filters::Subscriber`` must be initialized with the topic. And the ``tf2_ros::MessageFilter`` must be initialized with that ``Subscriber`` object.
+The other arguments of note in the ``MessageFilter`` constructor are the ``target_frame`` and the callback function. The target frame is the frame into which it will make sure ``canTransform`` will succeed. And the callback function is the function that will be called when the data is ready.
```
```suggestion
The other arguments of note in the ``MessageFilter`` constructor are the ``target_frame`` and the callback function.
The target frame is the frame into which it will make sure ``canTransform`` will succeed.
And the callback function is the function that will be called when the data is ready.
```

kennywangaigraphx replied:
I split this paragraph into 3 paragraphs.

clalancette commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,657 @@
+   }
+
+
+And lastly, the callback method. Once the data is ready, just call ``tf2_buffer_->transform`` and print outputs to the console.
```
```suggestion
And last, the callback method will call ``tf2_buffer_->transform`` when the data is ready and print output to the console.
```

kennywangaigraphx replied:
I changed these as your suggestion.

clalancette commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,657 @@
+   find_package(message_filters REQUIRED)
+   find_package(tf2_geometry_msgs REQUIRED)
+
+Most import, add these lines below the dependencies:
```
```suggestion
Most importantly, add these lines below the dependencies:
```

kennywangaigraphx replied:
I corrected this syntax error.

clalancette commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,657 @@
+
+   ros2 launch learning_tf2_py turtle_tf2_sensor_message.launch.py
+
+This will bring up the ``turtlesim`` window with two turtles, where ``turtle3`` is moving along a circle, while ``turtle1`` isn't moving at first. But you can run the ``turtle_teleop_key`` node in another terminal to drive ``turtle1`` to move:
```
```suggestion
This will bring up the ``turtlesim`` window with two turtles, where ``turtle3`` is moving along a circle, while ``turtle1`` isn't moving at first.
But you can run the ``turtle_teleop_key`` node in another terminal to drive ``turtle1`` to move:
```

kennywangaigraphx replied:
I split it into 2 paragraphs.

clalancette commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,657 @@
+
+   ros2 topic echo /turtle3/turtle_point_stamped
+
+Then there will have outputs like this:
```
```suggestion
Then there will be output like this:
```

kennywangaigraphx replied:
I corrected this error.

clalancette commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,657 @@
+
+   ros2 run learning_tf2_cpp turtle_tf2_message_filter
+
+If it's running right you should see streaming data like this:
```
```suggestion
If it's running correctly you should see streaming data like this:
```

kennywangaigraphx replied:
I changed that.

clalancette commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,657 @@
+Summary
+-------
+
+In this tutorial you learned how to use sensor data/messages in tf2. Specifically speaking, you learned how to publish ``PointStamped`` messages on a topic, and how to listen to the topic and transform the frame of ``PointStamped`` messages with ``tf2_ros::MessageFilter``.
```
```suggestion
In this tutorial you learned how to use sensor data/messages in tf2.
Specifically speaking, you learned how to publish ``PointStamped`` messages on a topic, and how to listen to the topic and transform the frame of ``PointStamped`` messages with ``tf2_ros::MessageFilter``.
```

kennywangaigraphx replied:
I split it into 2 paragraphs.

clalancette commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,665 @@
+2 Writing the message filter/listener node
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Now, to get the streaming ``PointStamped`` data of ``turtle3`` in the frame of ``turtle1`` reliably, we will create the source file of message filter/listener node at first.
```
```suggestion
Now, to get the streaming ``PointStamped`` data of ``turtle3`` in the frame of ``turtle1`` reliably, we will create the source file of the message filter/listener node.
```

kennywangaigraphx replied:
Ok, I changed that.

clalancette commented on source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst:
```diff
@@ -0,0 +1,665 @@
+   find_package(message_filters REQUIRED)
+   find_package(tf2_geometry_msgs REQUIRED)
+
+Most importly, add these lines below the dependencies:
```
```suggestion
Most importantly, add these lines below the dependencies:
```

kennywangaigraphx replied:
Yes, another error, thank you very much, I corrected it just now.

clalancette reviewed (changes requested):
This is looking really good now!  I've got two more minor edits, and then I think we are ready to merge.

clalancette reviewed (approved):
Looks good to me, thanks for the contribution and iterating.  @audrow @ahcorde Would one of you take a look and approve as well?

kennywangaigraphx commented:
@ahcorde  Would you review this tutorial for me when you are free? thank you very much.

ahcorde commented:
@Mergifyio backport foxy galactic
